tools/analyzers/nilvaluecheck/nilvaluecheck.go

```diff
@@ -0,0 +1,130 @@
+	skip := flagSet.String("skip-pkg", "", "package(s) to skip for linting")
+
+	return &analysis.Analyzer{
+		Name: "nilvaluecheck",
+		Doc:  "reports nil values used for specific types",
+		Run: func(pass *analysis.Pass) (interface{}, error) {
+			// Check for a skipped package.
+			skipped := strings.Split(*skip, ",")
+			for _, skipped := range skipped {
+				trimmed := strings.TrimSpace(skipped)
+				if strings.Contains(pass.Pkg.Path(), trimmed) {
+					return nil, nil
+				}
+			}
+
+			inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+
+			nodeFilter := []ast.Node{
+				(*ast.ReturnStmt)(nil),
+				(*ast.DeclStmt)(nil),
+			}
+
+			typePaths := strings.Split(*disalllowedPaths, ",")
+			hasTypePath := func(path string) bool {
+				for _, typePath := range typePaths {
+					trimmed := strings.TrimSpace(typePath)
+					if trimmed == path {
+						return true
+					}
+				}
+				return false
+			}
+
+			inspect.WithStack(nodeFilter, func(n ast.Node, push bool, stack []ast.Node) bool {
+				var funcType *ast.FuncType
+
+				if len(*disalllowedPaths) == 0 {
+					return false
+				}
+
+				for _, parent := range stack {
+					// Find the parent function declaration or function literal.
+					// Note that the stack orders from the top of the file, inward.
+					if funcDecl, ok := parent.(*ast.FuncDecl); ok {
+						funcType = funcDecl.Type
+					}
+
+					if r, ok := parent.(*ast.FuncLit); ok {
+						funcType = r.Type
+					}
+				}
+
+				if funcType == nil || funcType.Results == nil {
+					return false
+				}
+
+				switch s := n.(type) {
+				case *ast.ReturnStmt:
+					// Check for a return value of `nil` for any disallowed type.
+					for index, resultTypeExpr := range funcType.Results.List {
+						if resultTypeExpr.Type == nil {
+							return false
+						}
+
+						foundType := pass.TypesInfo.TypeOf(resultTypeExpr.Type)
+						if foundType == nil {
+							return false
+						}
+
+						resultType := foundType.String()
+						if hasTypePath(resultType) {
+							// Check for a nil result.
+							returnedValue := s.Results[index]
+							if ident, ok := returnedValue.(*ast.Ident); ok {
+								if ident.Name == "nil" {
+									pass.Reportf(s.Pos(), "In package %s: found `nil` returned for value of nil-disallowed type %s", pass.Pkg.Path(), resultType)
+								}
+							}
+						}
+					}
+
+				case *ast.DeclStmt:
+					// Check for a variable without an assigned value to a disallowed type.
+					if genDecl, ok := s.Decl.(*ast.GenDecl); ok && genDecl.Tok == token.VAR {
+						valueSpec := genDecl.Specs[0].(*ast.ValueSpec)
+						if valueSpec.Type == nil {
+							return false
+						}
+
+						foundType := pass.TypesInfo.TypeOf(valueSpec.Type)
+						if foundType == nil {
+							return false
+						}
+
+						varType := foundType.String()
+						if len(valueSpec.Values) == 0 && hasTypePath(varType) {
```

Comment:
looks like this catches `var thing *myType`, but not `var thing *mytype = nil` or `var thing *mytype = otherNilVariable`

Reply:
Correct. `var thing *mytype = nil` should be caught by existing lint and to do the `otherNilVariable` would require real flow analysis